src/actions/public/L2/estimateFees.ts

```diff
@@ -0,0 +1,65 @@
+import {
+  type Abi,
+  encodeFunctionData,
+  type EncodeFunctionDataParameters,
+  type EstimateGasParameters,
+  type PublicClient,
+  type Transport,
+} from 'viem'
+import { type Chain } from 'viem/chains'
+import type { GasPriceOracleParameters, OracleTransactionParameters } from '../../../types/gasPriceOracle.js'
+import { estimateL1Fee } from './estimateL1Fee.js'
+
+export type EstimateFeesParameters<
+  TAbi extends Abi | readonly unknown[],
+  TFunctionName extends string | undefined = undefined,
+> =
+  & OracleTransactionParameters<TAbi, TFunctionName>
+  & GasPriceOracleParameters
+  & Omit<EstimateGasParameters, 'data'>
+
+export type EstimateFees = <
+  TChain extends Chain | undefined,
+  TAbi extends Abi | readonly unknown[],
+  TFunctionName extends string | undefined = undefined,
+>(
+  client: PublicClient<Transport, TChain>,
+  options: EstimateFeesParameters<TAbi, TFunctionName>,
+) => Promise<bigint>
+
+/**
+ * Estimates gas for an L2 transaction including the l1 fee
+ * on non OP chains this is usually GasUsed * GasPrice
+ * on OP chains this is GasUsed * GasPrice + L1Fee
+ * @example
+ * const feeValue = await estimateFees(publicClient, {
+ *   abi,
+ *   functionName: balanceOf,
+ *   args: [address],
+ * });
+ */
+export const estimateFees: EstimateFees = async (client, options) => {
```

Comment:
in an ideal world, this would be the default `estimateGas` function for an OPStack chain in viem, right? @roninjin10 

Do we think we should name it like opStackEstimateGas or something closer?

Reply:
Estimate gas is different. On l1 estimate fees would be estimateGas multiplied by the gas price. This is something I chatted with Jake about and he lightly agreed it would be valuable to add for all chains

Reply:
ok thanks for clarifying @roninjin10

Reply:
I’m not in love with the name estimateFees btw I just couldn’t come up with a better name.  Etherscan calls it a transactionFee. Maybe we should call it transactionFee? I would love a name that makes it so folks can guess the difference between it and estimateGas most of the time

Reply:
transactionFee is pretty good! I don't have a strong preference